### PR TITLE
More comparisons

### PR DESCRIPTION
**Pull request by dellaert**
Added 
- Essential + K: uses new constructor to optimize for a single K
- "Binary + Ks" and "Binary + K": use `EssentialMatrixFactor`: makes error much less as only pairwise errors checked
It's clear that the binary `EssentialMatrixFactors` do much better in terms of geodesic distance, even without "enforcing consistency". They also optimize quite fast. Although: I have to admit I don't understand how they can do better than "Calibrated", which uses ground truth K. Maybe the transfer error pushes away from ground truth in an unexpected way.

image

### CODE REVIEW

File: gtsam/slam/EssentialMatrixFactor.h [hunk 1]
```diff
@@ -420,4 +417,104 @@ class EssentialMatrixFactor4
 };
 // EssentialMatrixFactor4
 
+/**
+ * Binary factor that optimizes for E and two calibrations Ka and Kb using the
+ * algebraic epipolar error (Ka^-1 pA)'E (Kb^-1 pB). The calibrations are
+ * assumed different for the two images. Don'tt use this factor with same
```
Comment (akshay-krishnan) [inline gtsam/slam/EssentialMatrixFactor.h]: nit: typo for `Don't`

File: gtsam/slam/EssentialMatrixFactor.h [hunk 2]
```diff
@@ -420,4 +417,104 @@ class EssentialMatrixFactor4
...
+ * calibration unknown, as Jacobians will be incorrect...
```
Comment (akshay-krishnan) [inline gtsam/slam/EssentialMatrixFactor.h]: can we check this? do you mean the case where keyKa == keyKb? we can maybe assert they're different in constructor?

Comment (akshay-krishnan) [reply gtsam/slam/EssentialMatrixFactor.h]: but if you mean "do not use this with different keys for calibrations that should be the same", yeah, that is a good comment to have.

File: python/gtsam/examples/ViewGraphComparison.py [hunk 3]
```diff
@@ -90,58 +90,80 @@ def compute_ground_truth(method, poses, cal):
...
     elif method == "Calibrated":
         FactorClass = gtsam.EssentialTransferFactorCal3f
-        # No priors on calibration needed
     else:
         raise ValueError(f"Unknown method {method}")
 
+    # Add priors on calibrations if necessary
+    if method in ["Essential+Ks", "Binary+Ks"]:
+        for i in range(num_cameras):
+            model = gtsam.noiseModel.Isotropic.Sigma(1, 1000.0)
```
Comment (akshay-krishnan) [inline python/gtsam/examples/ViewGraphComparison.py:120]: is that not a very high value? was there some rationale for it?

**Review by akshay-krishnan — APPROVED** [review]

File: gtsam/sfm/TransferFactor.h [hunk 4]
```diff
@@ -251,6 +253,24 @@ class EssentialTransferFactorK
...
+   * @param edge1 First EdgeKey specifying E1: (a, c) or (c, a)
+   * @param edge2 Second EdgeKey specifying E2: (b, c) or (c, b)
+   * @param keyK Calibration key for all views.
+   * @param triplets A vector of triplets containing (pa, pb, pc)
+   * @param model An optional SharedNoiseModel
+   */
+  EssentialTransferFactorK(EdgeKey edge1, EdgeKey edge2, Key keyK,
+                           const std::vector<Triplet>& triplets,
+                           const SharedNoiseModel& model = nullptr)
+      : Base(model, edge1, edge2, keyK, keyK, keyK),
```
Comment (travisdriver) [inline gtsam/sfm/TransferFactor.h:270]: nit: is it possible to change this so you don't have to send base 3 copies of the same key?

File: gtsam/slam/EssentialMatrixFactor.h [hunk 5]
```diff
@@ -420,4 +417,104 @@ class EssentialMatrixFactor4
...
+                       const CALIBRATION& Kb, OptionalMatrixType HE,
+                       OptionalMatrixType HKa,
+                       OptionalMatrixType HKb) const override {
+    // converting from pixel coordinates to normalized coordinates cA and cB
+    JacobianCalibration cA_H_Ka;  // dcA/dKa
+    JacobianCalibration cB_H_Kb;  // dcB/dKb
+    Point2 cA = Ka.calibrate(pA_, HKa ? &cA_H_Ka : 0, OptionalNone);
+    Point2 cB = Kb.calibrate(pB_, HKb ? &cB_H_Kb : 0, OptionalNone);
+
+    // convert to homogeneous coordinates
```
Comment (travisdriver) [inline gtsam/slam/EssentialMatrixFactor.h]: nit: capitalize first letter (and on other lines)

File: python/gtsam/examples/ViewGraphComparison.py (hunk 3 above)
Comment (travisdriver) [reply python/gtsam/examples/ViewGraphComparison.py:120]: Wondering the same

File: gtsam/slam/EssentialMatrixFactor.h [hunk 6]
```diff
@@ -420,4 +417,104 @@ class EssentialMatrixFactor4
 };
 // EssentialMatrixFactor4
 
+/**
+ * Binary factor that optimizes for E and two calibrations Ka and Kb using the
+ * algebraic epipolar error (Ka^-1 pA)'E (Kb^-1 pB). The calibrations are
+ * assumed different for the two images. Don'tt use this factor with same
+ * calibration unknown, as Jacobians will be incorrect...
+ *
+ * Note: even stronger caveats about having priors on calibration apply.
```
Comment (travisdriver) [inline gtsam/slam/EssentialMatrixFactor.h]: Could you elaborate on this?

File: gtsam/slam/EssentialMatrixFactor.h [hunk 7]
```diff
@@ -420,4 +417,104 @@ class EssentialMatrixFactor4
...
+
+  /**
+   *  Constructor
+   *  @param keyE Essential Matrix aEb variable key
+   *  @param keyKa Calibration variable key for camera A
+   *  @param keyKb Calibration variable key for camera B
+   *  @param pA point in first camera, in pixel coordinates
+   *  @param pB point in second camera, in pixel coordinates
+   *  @param model noise model is about dot product in ideal, homogeneous
+   * coordinates
```
Comment (travisdriver) [inline gtsam/slam/EssentialMatrixFactor.h:455]: nit: different indent

**Review by travisdriver — APPROVED** [review]
Looks great, excited to try out in GTSfM :)

File: gtsam/slam/EssentialMatrixFactor.h (hunk 1 above)
Comment (dellaert) [reply gtsam/slam/EssentialMatrixFactor.h]: Done

File: python/gtsam/examples/ViewGraphComparison.py (hunk 3 above)
Comment (dellaert) [reply python/gtsam/examples/ViewGraphComparison.py:120]: Did not want the optimization to diverge yet not overly influence result.

File: gtsam/sfm/TransferFactor.h (hunk 4 above)
Comment (dellaert) [reply gtsam/sfm/TransferFactor.h:270]: No, Base expects that.

File: gtsam/slam/EssentialMatrixFactor.h (hunk 2 above)
Comment (dellaert) [reply gtsam/slam/EssentialMatrixFactor.h]: I added a test asserting that 
```c++
    factor4.evaluateError(trueE, trueK, nullptr, &H0);
    factor5.evaluateError(trueE, trueK, trueK, nullptr, &H1, &H2);
    EXPECT(assert_equal(H0, H1 + H2, 1e-9));
```

File: gtsam/slam/EssentialMatrixFactor.h (hunk 5 above)
Comment (dellaert) [reply gtsam/slam/EssentialMatrixFactor.h]: Done

File: gtsam/slam/EssentialMatrixFactor.h (hunk 6 above)
Comment (dellaert) [reply gtsam/slam/EssentialMatrixFactor.h]: Done

File: gtsam/slam/EssentialMatrixFactor.h (hunk 7 above)
Comment (dellaert) [reply gtsam/slam/EssentialMatrixFactor.h:455]: Done

**Pull request merged**